Added authorization header option in OWL Ontology Loader

prakashsagadevan opened the pull request:
This commit added the authorization header in the OWL Ontology Loader of input stream from a URI.

prakashsagadevan commented:
@ignazio1977  can you please review the changes?

ansell commented on .travis.yml:
```diff
@@ -6,6 +6,4 @@ cache:
 language: java
 script: "mvn verify"
 jdk:
-  - openjdk7
```
Why were these Travis JDK's removed?

prakashsagadevan replied:
After merging the "owlapi_authorization" branch into "owlcs/owlapi:version4" branch, Maven pom file build plugin compiler source and target version are 1.8. so i have removed the jdk 1.7 reference.

Reference:
<plugin>
	<groupId>org.apache.maven.plugins</groupId>
	<artifactId>maven-compiler-plugin</artifactId>
	<version>3.6.0</version>
	<configuration>
		<source>1.8</source>
		<target>1.8</target>
		<encoding>UTF-8</encoding>
	</configuration>
</plugin>

Can you please let me know shall we add the  "openjdk7" and "oraclejdk7"?

ansell replied:
That looks okay to me. Just surprised that the .travis.yml file still contained JDK7 when it was compiled for JDK8.

ignazio1977 replied:
That's explained - the patch was originally for version 4.28, but version4 has moved to 4.5.0.

ansell commented on api/src/main/java/org/semanticweb/owlapi/io/AbstractOWLParser.java:
```diff
@@ -95,6 +95,9 @@ protected InputStream getInputStream(@Nonnull IRI documentIRI,
         String originalProtocol = originalURL.getProtocol();
         URLConnection conn = originalURL.openConnection();
         conn.addRequestProperty("Accept", actualAcceptHeaders);
+		if (config.isAcceptingAuthorization()) {
```
Tabs are not used by this codebase, please replace each tab with 4 spaces.

prakashsagadevan replied:
I have updated source and replaced each tab with 4 spaces.

ansell commented on api/src/main/java/org/semanticweb/owlapi/model/OWLOntologyLoaderConfiguration.java:
```diff
@@ -477,4 +489,35 @@ public OWLOntologyLoaderConfiguration setRepairIllegalPunnings(boolean b) {
+     *         to the new value.
+     */
+    @Nonnull
+    public OWLOntologyLoaderConfiguration setAcceptingAuthorization(boolean b) {
```
It doesn't make sense to have this value separate from the authorization header value itself. Could you remove this and integrate any functionality that relies on it with ``AUTHORIZATION_VALUE`` instead.

prakashsagadevan replied:
I have removed the ACCEPT_HTTP_AUTHORIZATION enum and updated the functionality that will check AuthorizationValue not null or empty before adding the Authorization header

ansell commented on api/src/main/java/org/semanticweb/owlapi/model/parameters/ConfigurationOptions.java:
```diff
@@ -97,7 +97,13 @@
+    REPAIR_ILLEGAL_PUNNINGS             (Boolean.TRUE),
+	/** True if http authorization 
+     * should be used. */
+    ACCEPT_HTTP_AUTHORIZATION             (Boolean.FALSE),
```
Remove this and replace any references to it with equivalent references to ``AUTHORIZATION_VALUE``

prakashsagadevan replied:
Removed  ACCEPT_HTTP_AUTHORIZATION and all the reference

ansell commented:
I will resolve the conflicts and merge this in. Thanks!

ignazio1977 commented:
I'm already working on that,
I.

ansell commented:
Ok

ignazio1977 commented:
Squashed and merged in commit ab9c80e691143a2046cbef330d2f563f1b2f6449 thanks for contributing

ansell commented:
Just one minor issue with ``Configuration.ACCEPT_HTTP_AUTHORIZATION`` that was removed in https://github.com/owlcs/owlapi/pull/716/commits/8eca08200f561b6a139464a62245ca1a9574a38c which came back to life in the squashed commit.

prakashsagadevan commented:
We have plans to implement the Basic Authentication for Protege desktop application. The Protege application uses the owlapi-api-4.2.8 jar file. So, we have updated the owlapi with the basic authentication support by adding the authorization header option. We had submitted our pull request (#716) and it has been merged into Version4 branch. 

Please find the details below.

Pull Request: Added authorization header option in OWL Ontology Loader #716
Pull Request URL: https://github.com/owlcs/owlapi/pull/716
Pull Request Fork Repository Branch : owlapi_authorization
Merged : version4

Can you please review the "prakashsagadevan:owlapi-4.2.9_add_authorization" branch and create a release tag for owlapi-parent-4.2.9? Or please let us know the procedure to create release tag 4.2.9.

Fork Repository: https://github.com/prakashsagadevan/owlapi.git
Branch Name: owlapi-4.2.9_add_authorization
Updated Version: 4.2.9

Additional question is, will this update be merged to all future releases of owlapi 5.x and 6.x?

ignazio1977 commented:
I'll do the 4.2.9 release later today. We'll port the fix to versions 5 and 6, of course :-)

ignazio1977 commented:
Having checked the 4.x releases and the version Protege trunk is using, I think we're missing something.

- The latest OWLAPI before changing to Java 8 is 4.3.2. Protege trunk is pointing at OWLAPI 4.3.1.
- The changes included in OWLAPI 4.3.2 and 4.5.0 were bug fixes and extensions to OWLAPI that the Protege developers were very interested in.
- The latest OWLAPI is intended to be a drop in replacement for Protege, meaning you could download the latest Protege build, replace the owlapi jar with the latest owlapi-osgidistribution from maven central, and everything should work as expected.
- I don't know the exact development plans for Protege, but I believe Protege will move officially to 4.5.0 or newer soon.

Therefore, it seems simpler and more useful to release version 4.5.1 instead of 4.2.9.

Would that cause a problem to your project, @prakashsagadevan ?

matthewhorridge commented:
> I don't know the exact development plans for Protege, but I believe Protege will move officially to 4.5.0 or newer soon.

Next release will be 4.5.0 (or any minor or point release from that)

ignazio1977 commented:
4.5.1 has been released and will be on maven central soon.

prakashsagadevan commented:
Hi @ignazio1977 

> The latest OWLAPI before changing to Java 8 is 4.3.2. Protege trunk is pointing at OWLAPI 4.3.1.
> The changes included in OWLAPI 4.3.2 and 4.5.0 were bug fixes and extensions to OWLAPI that the Protege developers were very interested in.

I think you are referring to the protege-owlapi (https://github.com/protegeproject/protege-owlapi/blob/master/pom.xml) github repository in this case. 

But I am looking to implement the Basic Authentication for Protege desktop application which is in (https://github.com/protegeproject/protege/blob/master/pom.xml) github repository. In this, Protege github repository master branch is pointing at OWLAPI 4.2.8 version.

**Reference :** https://mvnrepository.com/artifact/net.sourceforge.owlapi/owlapi-osgidistribution/4.2.8/usages

prakashsagadevan commented:
Hi @ignazio1977 

> The latest OWLAPI is intended to be a drop in replacement for Protege, meaning you could download the latest Protege build, replace the owlapi jar with the latest owlapi-osgidistribution from maven central, and everything should work as expected.

I have checked the owlapi 4.5.0 jar with latest Protege 5.2.0 application. But we are facing issues with Protege plugins.
Example, guava.jar file shows version incompatibility problems with Protege.

OWLAPI 4.5.0 requires the guava jar file dependency version as 20.0.

`<dependency>
 <groupId>com.google.guava</groupId>
 <artifactId>guava</artifactId>
 <version>20.0</version>
</dependency>`
But the Protege plugin requires the guava jar file dependency version as 18.0.

org.osgi.framework.BundleException: Unresolved constraint in bundle cellfie [22]: Unable to resolve 22.0: missing requirement [22.0] osgi.wiring.package; (&(osgi.wiring.package=com.google.common.base)(version>=18.0.0)(!(version>=19.0.0)))
 at org.apache.felix.framework.Felix.resolveBundleRevision(Felix.java:4002) ~[org.apache.felix.main.jar:na]
 at org.apache.felix.framework.Felix.startBundle(Felix.java:2045) ~[org.apache.felix.main.jar:na]
 at org.apache.felix.framework.Felix.setActiveStartLevel(Felix.java:1299) ~[org.apache.felix.main.jar:na]
 at org.apache.felix.framework.FrameworkStartLevelImpl.run(FrameworkStartLevelImpl.java:304) [org.apache.felix.main.jar:na]
 at java.lang.Thread.run(Unknown Source) [na:1.8.0_60]

**Reference:**
https://github.com/protegeproject/protege/pull/633
https://github.com/Galigator/openllet/issues/18

prakashsagadevan commented:
Hi @ignazio1977 

> I don't know the exact development plans for Protege, but I believe Protege will move officially to 4.5.0 or newer soon.

Protege desktop application's (https://github.com/protegeproject/protege/blob/master/pom.xml) master github repository is pointing to OWLAPI 4.2.8 version. We will check with Protege development team for their upgrade plans.

Meanwhile, is it possible for you to release the owlapi-parent-4.2.9? Thanks much in advance.

prakashsagadevan commented:
Hi @matthewhorridge 

> Next release will be 4.5.0 (or any minor or point release from that)

Please let me know if we have plans to include OWLAPI 4.5.0 in the Protege desktop application (https://github.com/protegeproject/protege/blob/master/pom.xml) github repository?

ignazio1977 commented:
> Meanwhile, is it possible for you to release the owlapi-parent-4.2.9? Thanks much in advance.

Done.

prakashsagadevan commented:
-:) Thanks for update. Please let me know will the jar file available in mvn repository.

ignazio1977 commented:
That's what I meant when I said done :-)

prakashsagadevan commented:
Still, I am not able to find the OWL API 4.2.9 & 4.5.1 jar files in the maven repository. Can you please check the below maven repository URL?
https://mvnrepository.com/artifact/net.sourceforge.owlapi/owlapi-osgidistribution

ignazio1977 commented:
Check Maven Central. I don't know the update policies for mvnrepository.com

http://search.maven.org/#search%7Cgav%7C1%7Cg%3A%22net.sourceforge.owlapi%22%20AND%20a%3A%22owlapi-osgidistribution%22

prakashsagadevan commented:
I am able to download the OWL API 4.2.9 jar from the maven repository with below dependency configuration. 
```
<dependency>
	<groupId>net.sourceforge.owlapi</groupId>
	<artifactId>owlapi-osgidistribution</artifactId>
	<version>4.2.9</version>
</dependency>
```
Thanks for your help  :+1:
